locations/spiders/engen.py

```diff
@@ -1,23 +1,44 @@
-import scrapy
+from locations.categories import Access, Categories, Fuel, apply_category, apply_yes_no
+from locations.json_blob_spider import JSONBlobSpider
 
-from locations.categories import Categories, Fuel, apply_yes_no
-from locations.dict_parser import DictParser
 
-
-class EngenSpider(scrapy.Spider):
+class EngenSpider(JSONBlobSpider):
     name = "engen"
-    item_attributes = {"brand": "Engen", "brand_wikidata": "Q3054251", "extras": Categories.FUEL_STATION.value}
-    allowed_domains = ["engen.co.za"]
+    item_attributes = {"brand": "Engen", "brand_wikidata": "Q3054251"}
     start_urls = ["https://engen-admin.engen.co.za/api/service-stations/all"]
+    locations_key = ["response", "data", "stations"]
 
-    def parse(self, response):
-        data = response.json()
-        for i in data["response"]["data"]["stations"]:
-            i["branch"] = i.pop("company_name").replace(self.item_attributes["brand"], "").strip()
-            i["phone"] = i.pop("mobile")
-            postcode = i.pop("street_postal_code")
-            if postcode and postcode != "0":
-                i["postcode"] = postcode
-            item = DictParser.parse(i)
-            apply_yes_no(Fuel.ADBLUE, item, "AdBlue" in i["rental_units"])
-            yield item
+    def post_process_item(self, item, response, location):
+        item["branch"] = location.pop("company_name").replace(self.item_attributes["brand"], "").strip()
+        item["phone"] = location.pop("mobile")
+        postcode = location.pop("street_postal_code")
+        if postcode and postcode != "0":
+            location["postcode"] = postcode
+        item["street_address"] = item.pop("street")
+        try:
+            int(item["street_address"].split(" ", 1)[0])
+            item["housenumber"] = item["street_address"].split(" ", 1)[0]
+            item["street"] = item["street_address"].split(" ", 1)[1]
+        except ValueError:
+            pass
+        if "Quickshop" in location.get("rental_units", ""):
+            shop_item = item.deepcopy()
+            shop_item["ref"] = str(item.get("ref")) + "-attached-shop"
+            shop_item.pop("email")
+            shop_item.pop("phone")
+            shop_item.update(
+                {
+                    "brand": "Quickshop",
+                    "brand_wikidata": "Q122764368",
+                    "extras": Categories.SHOP_CONVENIENCE.value,
```

Comment:
Should use `apply_category`. This will nuke any values already in `extras`.

Reply:
Thanks